lib/spack/spack/spec_list.py

```diff
@@ -11,19 +11,6 @@
 from spack.spec import Spec
 
 
-def spec_ordering_key(s):
-    if s.startswith('^'):
-        return 5
-    elif s.startswith('/'):
-        return 4
-    elif s.startswith('%'):
-        return 3
-    elif any(s.startswith(c) for c in '~-+@') or '=' in s:
-        return 2
-    else:
-        return 1
```

Comment:
This removed function was trying to account for associativity when using string concatenation

Reply:
So I guess the idea was already to have the `Spec.constrain` semantics, but the implementation was failing for anonymous specs which had a constraint both on the root and on a dependency